rslib/src/import_export/gather.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::{
+    collections::{HashMap, HashSet},
+    path::PathBuf,
+};
+
+use itertools::Itertools;
+
+use crate::{
+    decks::NormalDeck,
+    io::filename_is_safe,
+    latex::extract_latex,
+    prelude::*,
+    revlog::RevlogEntry,
+    text::{extract_media_refs, extract_underscored_css_imports, extract_underscored_references},
+};
+
+#[derive(Debug, Default)]
+pub(super) struct ExportData {
+    pub(super) decks: Vec<Deck>,
+    pub(super) notes: Vec<Note>,
+    pub(super) cards: Vec<Card>,
+    pub(super) notetypes: Vec<Notetype>,
+    pub(super) revlog: Vec<RevlogEntry>,
+    pub(super) decks_configs: Vec<DeckConfig>,
+    pub(super) media_paths: HashSet<PathBuf>,
+}
+
+impl ExportData {
+    pub(super) fn gather_data(
+        &mut self,
+        col: &mut Collection,
+        search: impl TryIntoSearch,
+        with_scheduling: bool,
+    ) -> Result<()> {
+        self.notes = col.gather_notes(search)?;
+        self.cards = col.gather_cards()?;
+        self.decks = col.gather_decks()?;
+        self.notetypes = col.gather_notetypes()?;
+
+        if with_scheduling {
+            self.revlog = col.gather_revlog()?;
+            self.decks_configs = col.gather_deck_configs(&self.decks)?;
+        } else {
+            self.remove_scheduling_information(col);
+        };
+
+        col.storage.clear_searched_notes_table()?;
+        col.storage.clear_searched_cards_table()
+    }
+
+    pub(super) fn gather_media_paths(&mut self) {
+        let mut inserter = |name: &str| {
+            if filename_is_safe(name) {
+                self.media_paths.insert(PathBuf::from(name));
+            }
+        };
+        let svg_getter = svg_getter(&self.notetypes);
+        for note in self.notes.iter() {
+            gather_media_paths_from_note(note, &mut inserter, &svg_getter);
+        }
+        for notetype in self.notetypes.iter() {
+            gather_media_paths_from_notetype(notetype, &mut inserter);
+        }
+    }
+
+    fn remove_scheduling_information(&mut self, col: &Collection) {
+        self.remove_system_tags();
+        self.reset_deck_config_ids();
+        self.reset_cards(col);
+    }
+
+    fn remove_system_tags(&mut self) {
+        // TODO: case folding? child tags?
```

Comment:
Anki assigns these tags with consistent lower case. Should I still match unicase? And what if a user used them as parent tags?

Reply:
They're ascii and not localized, so to_ascii_lowercase() should suffice. I don't think I've ever seen subtags used here, so wouldn't worry about it.